libkbfs/rekey_queue.go

```diff
@@ -37,148 +60,114 @@ var _ RekeyQueue = (*RekeyQueueStandard)(nil)
 	rkq.log.Debug("Enqueueing %s for rekey", id)
-	c := make(chan error, 1)
-	rkq.wg.Add(1)
-	func() {
-		rkq.queueMu.Lock()
-		defer rkq.queueMu.Unlock()
-		if rkq.cancel == nil {
-			// create a new channel
-			rkq.hasWorkCh = make(chan struct{}, 1)
-			// spawn goroutine if needed
-			var ctx context.Context
-			ctx, rkq.cancel = context.WithCancel(context.Background())
-			go rkq.processRekeys(ctx, rkq.hasWorkCh)
-		}
-		rkq.queue = append(rkq.queue, rekeyQueueEntry{id, c})
-	}()
-	// poke the channel
-	select {
-	case rkq.hasWorkCh <- struct{}{}:
-	default:
+
+	if ch := rkq.GetRekeyChannel(id); ch != nil {
+		return ch
 	}
-	return c
+
+	rkq.wg.Add(1)
+
+	rkq.lock.Lock()
+	defer rkq.lock.Unlock()
+
+	rkq.ensureRunningLocked()
+
+	ch := make(chan error, 1)
+	rkq.queue <- rekeyQueueEntry{id: id, ch: ch}
+	rkq.pendings[id] = ch
+
+	return ch
 }
 
 // IsRekeyPending implements the RekeyQueue interface for RekeyQueueStandard.
 func (rkq *RekeyQueueStandard) IsRekeyPending(id tlf.ID) bool {
-	return rkq.GetRekeyChannel(id) != nil
+	rkq.lock.RLock()
+	defer rkq.lock.RUnlock()
+	_, ok := rkq.pendings[id]
+	return ok
 }
 
 // GetRekeyChannel implements the RekeyQueue interface for RekeyQueueStandard.
 func (rkq *RekeyQueueStandard) GetRekeyChannel(id tlf.ID) <-chan error {
-	rkq.queueMu.RLock()
-	defer rkq.queueMu.RUnlock()
-	for _, e := range rkq.queue {
-		if e.id == id {
-			return e.ch
-		}
+	rkq.lock.RLock()
+	defer rkq.lock.RUnlock()
+	if ch, ok := rkq.pendings[id]; ok {
+		return ch
 	}
 	return nil
 }
 
 // Clear implements the RekeyQueue interface for RekeyQueueStandard.
 func (rkq *RekeyQueueStandard) Clear() {
-	channels := func() []chan error {
-		rkq.queueMu.Lock()
-		defer rkq.queueMu.Unlock()
-		if rkq.cancel != nil {
-			// cancel
-			rkq.cancel()
-			rkq.cancel = nil
-		}
-		// collect channels and clear queue
-		var channels []chan error
-		for _, e := range rkq.queue {
-			channels = append(channels, e.ch)
-		}
-		rkq.queue = make([]rekeyQueueEntry, 0)
-		return channels
-	}()
-	for _, c := range channels {
-		c <- context.Canceled
-		close(c)
+	rkq.lock.Lock()
+	defer rkq.lock.Unlock()
+	if rkq.cancel == nil {
+		return
 	}
-}
-
-// Wait implements the RekeyQueue interface for RekeyQueueStandard.
-func (rkq *RekeyQueueStandard) Wait(ctx context.Context) error {
-	return rkq.wg.Wait(ctx)
-}
-
-// CtxRekeyTagKey is the type used for unique context tags within an
-// enqueued Rekey.
-type CtxRekeyTagKey int
 
-const (
-	// CtxRekeyIDKey is the type of the tag for unique operation IDs
-	// within an enqueued Rekey.
-	CtxRekeyIDKey CtxRekeyTagKey = iota
-)
-
-// CtxRekeyOpID is the display name for the unique operation
-// enqueued rekey ID tag.
-const CtxRekeyOpID = "REKEYID"
-
-// Dedicated goroutine to process the rekey queue.
-func (rkq *RekeyQueueStandard) processRekeys(ctx context.Context, hasWorkCh chan struct{}) {
-	for {
+	rkq.cancel()
+	for more := true; more; { // drain rkq.queue
```

Comment:
hmm... doesn't it block when `rkq.queue` is empty but not closed? This approach would drain it and move on when it's empty